### PR TITLE
Multisig: adds expiration to proposals

### LINKED ISSUE #3: Multisig date expiration for proposals
We need to extend the proposal functionality and allow the creator of proposals to set a deadline for them

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/multisig/src/contract.rs [hunk 1]
```diff
@@ -137,12 +138,19 @@ impl Multisig {
             title: title.clone(),
             description,
         };
+
+        let creation_timestamp = env.ledger().timestamp();
+        let expiration_timestamp = creation_timestamp + deadline.unwrap_or(SEVEN_DAYS_DEADLINE);
```
Comment (ueco-jb) [inline contracts/multisig/src/contract.rs]: Add some check that the deadline is not earlier or equal to the time right now, it must be longer then 1 hour (just for the sake if someone doesn't get timestamp format and puts 1 or something). Return an error if not.

File: contracts/multisig/src/contract.rs [hunk 2]
```diff
@@ -166,11 +179,16 @@ impl Multisig {
         }
 
         let proposal_id = increment_last_proposal_id(&env);
+        let creation_timestamp = env.ledger().timestamp();
+        let expiration_timestamp = creation_timestamp + deadline.unwrap_or(SEVEN_DAYS_DEADLINE);
+
```
Comment (ueco-jb) [inline contracts/multisig/src/contract.rs]: Checks are missing like mentioned above, and here if `deadline` is not present just don't modify the value.

File: contracts/multisig/src/contract.rs [hunk 3]
```diff
@@ -211,6 +229,15 @@ impl Multisig {
...
 
+        let curr_timestamp = env.ledger().timestamp();
+        if curr_timestamp > proposal.expiration_timestamp {
+            log!(
+                &env,
+                "Multisig: Sign proposal: Trying to sign an expired proposal!"
+            );
+            panic_with_error!(&env, ContractError::ProposalExpired);
+        }
+
```
Comment (ueco-jb) [inline contracts/multisig/src/contract.rs]: Although pointless, signing is fine.

```suggestion
```

File: contracts/multisig/src/contract.rs [hunk 4]
```diff
@@ -251,6 +278,15 @@ impl Multisig {
...
 
+        let curr_timestamp = env.ledger().timestamp();
+        if curr_timestamp > proposal.expiration_timestamp {
+            log!(
+                &env,
+                "Multisig: Execute proposal: Trying to execute an expired proposal!"
+            );
+            panic_with_error!(&env, ContractError::ProposalExpired);
+        }
+
```
Comment (ueco-jb) [inline contracts/multisig/src/contract.rs]: At this point we could also automatically set `proposal_state.ProposalExpired`.

**Review by ueco-jb — COMMENTED** [review]
Good work, some comments.

File: contracts/multisig/src/storage.rs [hunk 5]
```diff
@@ -7,6 +7,9 @@ pub struct Proposal {
     pub sender: Address,
     pub proposal: ProposalType,
     pub status: ProposalStatus,
+    // NOTE: we can easily go without this if we wanted to
```
Comment (ueco-jb) [inline contracts/multisig/src/storage.rs]:
```suggestion
```

File: contracts/multisig/src/contract.rs (hunk 1 above)
Comment (gangov) [reply contracts/multisig/src/contract.rs]: `expiration_date` wlil always be over `creation_timestamp`, so I'm checking if we are at least an hour on top of that.

File: contracts/multisig/src/contract.rs (hunk 2 above)
Comment (gangov) [reply contracts/multisig/src/contract.rs]: I added the same check as in `create_transaction_proposal`

File: contracts/multisig/src/contract.rs (hunk 4 above)
Comment (gangov) [reply contracts/multisig/src/contract.rs]: added that to the source, but I don't think we can test it currently unless we change the return type to `Result`

File: contracts/multisig/src/contract.rs (hunk 1 above)
Comment (ueco-jb) [reply contracts/multisig/src/contract.rs]: But this is wrong then, it should be
```suggestion
        let expiration_timestamp = deadline.unwrap_or(creation_timestamp +SEVEN_DAYS_DEADLINE);
        if expiration_timestamp <= creation_timestamp {
            return Err()
```
And rename this  parameter from `deadline` to `expiration_date`.

**Review by ueco-jb — COMMENTED** [review]
You didn't actually test the most basic scenario - someone provides an expiration timestamp, time moves within that date and the proposal is executed. That would be the typical use case.

**Comment by gangov** [comment]
> You didn't actually test the most basic scenario - someone provides an expiration timestamp, time moves within that date and the proposal is executed. That would be the typical use case.

I'll add that

**Review by ueco-jb — APPROVED** [review]
LGTM

**Pull request merged**